### PR TITLE
[Require] Ensure -I beats a default gem

### LINKED ISSUE #1866: -I should take precedence over default gems
I'm having a problem or would like to suggest a feature.

My current problem is that the following test fails by activating `c`

```diff
diff --git a/test/rubygems/test_require.rb b/test/rubygems/test_require.rb
index dd606e44..63bc3461 100644
--- a/test/rubygems/test_require.rb
+++ b/test/rubygems/test_require.rb
@@ -80,6 +80,37 @@ class TestGemRequire < Gem::TestCase
     Object.send :remove_const, :HELLO if Object.const_defined? :HELLO
   end
 
+  # Providing -I on the commandline should always beat gems
+  def test_dash_i_beats_default_gems
+    a1 = new_default_spec "a", "1", {"b" => "= 1"}, "test_gem_require_a.rb"
+    b1 = new_default_spec "b", "1", {"c" => "> 0"}, "b/c.rb"
+    c1 = new_default_spec "c", "1", nil, "c/c.rb"
+    c2 = new_default_spec "c", "2", nil, "c/c.rb"
+
+    install_default_specs c1, c2, b1, a1
+
+    dir = Dir.mktmpdir("test_require", @tempdir)
+    dash_i_arg = File.join dir, 'lib'
+
+    c_rb = File.join dash_i_arg, 'b', 'c.rb'
+
+    FileUtils.mkdir_p File.dirname c_rb
+    File.open(c_rb, 'w') { |f| f.write "class Object; HELLO = 'world' end" }
+
+    lp = $LOAD_PATH.dup
+
+    # Pretend to provide a commandline argument that overrides a file in gem b
+    $LOAD_PATH.unshift dash_i_arg
+
+    assert_require 'test_gem_require_a'
+    assert_require 'b/c' # this should be required from -I
+    assert_equal "world", ::Object::HELLO
+    assert_equal %w(a-1), loaded_spec_names
+  ensure
+    $LOAD_PATH.replace lp
+    Object.send :remove_const, :HELLO if Object.const_defined? :HELLO
+  end
+
   def test_concurrent_require
     skip 'deadlock' if /^1\.8\./ =~ RUBY_VERSION
 
```

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [ ] The command line `gem`
  - [x] Other

Here are my current environment details:

```
$ gem env version
2.6.10
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by segiddins**
# Description:

Fixes #1866.
Definitely a WIP, just wanted to let everyone have a look at this.

# Tasks:

- [ ] Describe the problem / feature
- [ ] Write tests
- [ ] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

File: lib/rubygems.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@
 # All rights reserved.
 # See LICENSE.txt for permissions.
 #++
-
+require 'irb'
```
Comment (segiddins) [inline lib/rubygems.rb]: remove

File: lib/rubygems/core_ext/kernel_require.rb [hunk 2]
```diff
@@ -41,6 +41,35 @@ def require path
...
+          full_path = File.join(lp, "#{path}#{s}")
+          if File.file?(full_path)
+            rp = full_path
+            break
+          end
+        end
+        break if rp
+      end
+      rp
+    end
```
Comment (duckinator) [inline lib/rubygems/core_ext/kernel_require.rb]: Maybe rearrange to look something like this?

```ruby
resolved_path =
  if path.start_with?('/')
    path
  else
    <snip>
  end
```

File: lib/rubygems/core_ext/kernel_require.rb [hunk 3]
```diff
@@ -41,6 +41,35 @@ def require path
...
+      return if $LOADED_FEATURES.include?(resolved_path)
+      if spec = Gem.find_unresolved_default_spec(resolved_path)
+        Gem.remove_unresolved_default_spec(spec)
+        Kernel.send(:gem, spec.name)
+      else
+        # p "unable to find default spec #{path} w/ rp #{resolved_path}"
+      end
+      RUBYGEMS_ACTIVATION_MONITOR.exit
+      return gem_original_require(path)
+    end
```
Comment (duckinator) [inline lib/rubygems/core_ext/kernel_require.rb]: Whitespace can greatly readability. Maybe something more like this?

```ruby
    if resolved_path
      return if $LOADED_FEATURES.include?(resolved_path)

      if spec = Gem.find_unresolved_default_spec(resolved_path)
        Gem.remove_unresolved_default_spec(spec)
        Kernel.send(:gem, spec.name)
      else
        # p "unable to find default spec #{path} w/ rp #{resolved_path}"
      end

      RUBYGEMS_ACTIVATION_MONITOR.exit

      return gem_original_require(path)
    end
```

**Comment by duckinator** [comment]
I know you said this is a WIP, but I left some comments with formatting suggestions anyway. :smile:

**Comment by segiddins** [comment]
I need suggestions on the fundamentals of the change more than the formatting

**Comment by indirect** [comment]
@segiddins to be fair, this is one of those "unclear on what the potential problems are" situations. :/ what tests does this break? why does it break those tests? should we then think about making this a breaking change? is there another way to implement this behavior that wouldn't be a breaking change? it seems pretty arguable that it was a bug that default gems ever won against -I.

**Comment by segiddins** [comment]
Those are the things I need help with 😝

**Comment by indirect** [comment]
@segiddins let's talk about it, then! Write up your understanding of the problem, the current tests, and any theories you have about why this is breaking tests. Then tell us your ideas about what to do, and we can discuss. :)

**Comment by bronzdoc** [comment]
Is this good to go @deivid-rodriguez ?

**Comment by deivid-rodriguez** [comment]
I had rebased and fix CI, but I was wondering if this could affect `require` performance and wanted to have another look. Also, I saw something new in core that I think could help here: https://github.com/ruby/ruby/commit/4a935bc6f66fda448fb430c7627a2f175dcd0650#diff-690d269e57db631439f0d63ee2dbe9e6.

**Comment by deivid-rodriguez** [comment]
I wrote a script to measure how the performance of `require` is affected by this patch.

This is the `bench.rb` script I wrote (to be run from a rubygems clone).

```ruby
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("lib")

require "rubygems"
require "benchmark/ips"
require "fileutils"

(1..10000).each do |d|
  FileUtils.mkdir_p("tmp/#{d}")
end

[1, 10, 100, 1000, 10000].each do |load_path_size|
  puts "Benchmark for LOAD_PATH size #{load_path_size}..."

  (0..load_path_size).to_a.reverse.each do |d|
    expanded_path = File.expand_path("tmp/#{d}")
    $LOAD_PATH.unshift(expanded_path) unless $LOAD_PATH.include?(expanded_path)
  end

  Benchmark.ips do |x|
    x.report("regular_mode") do |times|
      ENV["RG_LOAD_PATH_PRIORITY"] = nil

      i = 0

      while i < times
        # require "active_support/core_ext/kernel/singleton_class"
        begin
          require "superkalifragislisticoespialidoso"
        rescue LoadError
        end

        $LOADED_FEATURES.pop

        i += 1

      end
    end

    x.report("new_mode") do |times|
      ENV["RG_LOAD_PATH_PRIORITY"] = "true"

      i = 0

      while i < times
        # require "active_support/core_ext/kernel/singleton_class"
        begin
          require "superkalifragislisticoespialidoso"
        rescue LoadError
        end

        $LOADED_FEATURES.pop

        i += 1
      end
    end

    x.compare!
  end
end
```

The results on my system are

```
$ ruby --disable-gems bench.rb
Benchmark for LOAD_PATH size 1...
Warming up --------------------------------------
        regular_mode   808.000  i/100ms
            new_mode   446.000  i/100ms
Calculating -------------------------------------
        regular_mode      7.367k (±11.8%) i/s -     36.360k in   5.019450s
            new_mode      4.487k (±10.7%) i/s -     22.300k in   5.043096s

Comparison:
        regular_mode:     7366.7 i/s
            new_mode:     4486.8 i/s - 1.64x  slower

Benchmark for LOAD_PATH size 10...
Warming up --------------------------------------
        regular_mode   549.000  i/100ms
            new_mode   217.000  i/100ms
Calculating -------------------------------------
        regular_mode      5.087k (± 9.2%) i/s -     25.254k in   5.008382s
            new_mode      2.357k (± 6.8%) i/s -     11.935k in   5.089742s

Comparison:
        regular_mode:     5086.8 i/s
            new_mode:     2357.5 i/s - 2.16x  slower

Benchmark for LOAD_PATH size 100...
Warming up --------------------------------------
        regular_mode   137.000  i/100ms
            new_mode    39.000  i/100ms
Calculating -------------------------------------
        regular_mode      1.325k (± 6.0%) i/s -      6.713k in   5.084980s
            new_mode    400.783  (± 6.7%) i/s -      2.028k in   5.084453s

Comparison:
        regular_mode:     1325.0 i/s
            new_mode:      400.8 i/s - 3.31x  slower

Benchmark for LOAD_PATH size 1000...
Warming up --------------------------------------
        regular_mode    12.000  i/100ms
            new_mode     4.000  i/100ms
Calculating -------------------------------------
        regular_mode    139.879  (±11.4%) i/s -    696.000  in   5.042779s
            new_mode     39.973  (±10.0%) i/s -    200.000  in   5.058432s

Comparison:
        regular_mode:      139.9 i/s
            new_mode:       40.0 i/s - 3.50x  slower

Benchmark for LOAD_PATH size 10000...
Warming up --------------------------------------
        regular_mode     1.000  i/100ms
            new_mode     1.000  i/100ms
Calculating -------------------------------------
        regular_mode     13.645  (± 7.3%) i/s -     68.000  in   5.041690s
            new_mode      4.025  (± 0.0%) i/s -     21.000  in   5.229632s

Comparison:
        regular_mode:       13.6 i/s
            new_mode:        4.0 i/s - 3.39x  slower

```


I think we could introduce this fix without overhead using the new `$LOAD_PATH.resolve_feature_path(feature)` method that will be introduced in ruby 2.7 but I'm afraid we would need to request it to be expanded to allow ignoring the builtin LOAD_PATH, something like
`$LOAD_PATH.resolve_feature_path(feature, ignore_builtin_entries: false)`.

But I'm not sure if ruby-core will be willing to introduce this... :grimacing:.

**Comment by deivid-rodriguez** [comment]
Another option is to consider that this patch is probably fine because applications or libraries which big LOAD_PATHs are most likely using `bundler`, which does not monkeypatch `require`, so in those cases, the only affected requires would be those _before_ `require "bundler/setup"`, probably only a few.

@hsbt What do you think?

**Comment by hsbt** [comment]
@deivid-rodriguez I just reviewed this. I understood this feature and concerns about performance. I tried to count `$LOAD_PATH` the real-world application used redmine.

https://gist.github.com/hsbt/86870742cb4f0424ec64b17fcec9086f

It returns 120+ paths included system paths. I think the dependencies our production system was larger than redmine. 

I bit of worries performance regression. But We tried to merge this and evaluate performance issue with Ruby 2.7.0-preview2 before the final release.

**Comment by deivid-rodriguez** [comment]
Thanks for having a look @hsbt. Does you realworld application use bundler? Note that `bundler` does not use a monkeypatched `require`. See: https://github.com/bundler/bundler/blob/421ceeca55871e8887f3290325078b5946798aa5/lib/bundler/rubygems_integration.rb#L303-L312.

So this patch will only affect performance of any requires _before_ `bundler/setup` is required, probably not many.

**Comment by deivid-rodriguez** [comment]
There's also https://github.com/rubygems/rubygems/pull/2741, that should slightly improve the `require` performance, but only for default gems.

**Comment by hsbt** [comment]
>So this patch will only affect performance of any requires before bundler/setup is required, probably not many.

Ah, I got it. That being the case, This performance behavior is a little issue for production system especially Rails application.

**Comment by deivid-rodriguez** [comment]
Ok, let's do this then, so we can unblock the rest of PRs depending on this!

@bundlerbot r=hsbt

**Comment by hsbt** [comment]
@nobu I will remove `-1` index in ruby/ruby.

**Comment by deivid-rodriguez** [comment]
I'm not sure... The most likely explanation is that since this PR is very old, maybe when I first rebased it, we still supported MRI 1.8. Thanks for removing the unneeded fallback!

**Comment by segiddins** [comment]
Yay thanks for landing this!

**Pull request merged**